README.md
sftp cloudfs
============

This is a SFTP (Secure File Transfer Protocol) interface to OpenStack
Object Storage, providing a service that acts as a proxy between a
SFTP client and a storage service.

The username/password pair used to open the SFTP session is validated
using the authentication service of the files/storage service to get
an authentication token.

The communication between the client and the SFTP daemon is encrypted
all the time, and the SFTP service supports HTTPS communication with
the remote files/storage service.

There's limited SCP support since 0.10.


Install
-------

Requirements:

- python (2.6)
- paramiko (1.7.6+; 1.17.0+ recommended)
- python-swiftclient (2.0+)
- python-daemon (1.5.5)
- ftp-cloudfs (0.35+)
- python-memcached (1.45)

These are the minimum recommended versions based in our testing
environment.

You may need to create a host key with ssh-keygen.

To install the software, run following command:

    python setup.py install

Or using pip:

    pip install sftp-cloudfs

Please use the latest pip version, old versions may have bugs. You
can upgrade pip using pip: pip install --upgrade pip.


Usage
-----

Once installed you can run the service with sftpcloudfs executable,
that supports following options:

    --version             show program's version number and exit
    -h, --help            show this help message and exit
    -a AUTHURL, --auth-url=AUTHURL
                          Authentication URL
    --insecure            Allow to access servers without checking SSL certs
    -k HOST_KEY, --host-key-file=HOST_KEY
                          Host RSA key used by the server
    -b BIND_ADDRESS, --bind-address=BIND_ADDRESS
                          Address to bind (default: 127.0.0.1)
    -p PORT, --port=PORT  Port to bind (default: 8022)
    --server-ident=SERVER_IDENT
                          Server ident to use when sending the SSH banner to the
                          client (default: sftpcloudfs_VER)
    --memcache=MEMCACHE   Memcache server(s) to be used for cache (ip:port)
    -l LOG_FILE, --log-file=LOG_FILE
                          Log into provided file
    -f, --foreground      Run in the foreground (don't detach from terminal)
    --disable-scp         Disable SCP support (default: enabled)
    --syslog              Enable logging to system logger (daemon facility)
    -v, --verbose         Show detailed information on logging
    --pid-file=PID_FILE   Full path to the pid file location
    --uid=UID/USER        UID or USER to drop the privileges to when in daemon mode
    --gid=GID/GROUP       GID or GROUP to drop the privileges to when in daemon mode
    --keystone-auth       Use OpenStack Identity Service (Keystone, requires keystoneclient)
    --keystone-auth-version=VERSION
                          Identity API version to be used, default: 2.0
    --keystone-region-name=REGION_NAME
                          Region name to be used in Keystone auth
    --keystone-tenant-separator=TENANT_SEPARATOR
                          Character used to separate tenant_name/username in
                          Keystone auth, default: TENANT.USERNAME
    --keystone-domain-separator=DOMAIN_SEPARATOR
                          Character used to separate project_name/project_domain_name
                          and username/user_domain_name in Keystone auth v3, default: @
    --keystone-service-type=SERVICE_TYPE
                          Service type to be used in Keystone auth,
                          default: object-store
    --keystone-endpoint-type=ENDPOINT_TYPE
                          Endpoint type to be used in Keystone auth,
                          default: publicURL
    --storage-policy=STORAGE_POLICY
                          Swift storage policy to be used (optional)
                          Access to other containers will be denied
    --config=CONFIG       Use an alternative configuration file

The default location for the configuration file is /etc/sftpcloudfs.conf.

Memcache is optional but highly recommended for better performance. Any Memcache
server must be secured to prevent unauthorized access to the cached data.

By default Swift auth 1.0 will be used, and is compatible with OpenStack
Object Storage (Swift) using swauth authentication middleware.

Optionally OpenStack Identity Service (*aka* keystone) v2.0 or v3 can be used.
Currently python-keystoneclient (0.3.2+ recommended) is required to use Keystone auth
and it can be enabled with ``keystone-auth`` option.

The server supports large files (over the 5GB default) by splitting the files
in parts into a *.part* subdirectory and using a manifest file to access them as
a single file.

With storage-policy parameter, you can restrict user access to a single policy.
If no name is specified, the default policy is used (and if no other policies, defined
Policy-0 is considered the default).

Please check the example configuration file for further details.


License
-------

This is free software under the terms of MIT license (check COPYING file
included in this package).

The server is loosely based on the BSD licensed sftpd server code from:

    http://code.google.com/p/pyfilesystem/


Contact and support
-------------------

The project website is at:

  https://github.com/memset/sftpcloudfs

There you can file bug reports, ask for help or contribute patches.


Authors
-------

- Nick Craig-Wood <nick@memset.com>
- Juan J. Martinez <jjm@usebox.net>

Contributors
------------

- Christophe Le Guern <c35sys@gmail.com>
- Koert van der Veer <koert@cloudvps.com>
- Julien Lutran <julien.lutran@ovh.net>



ChangeLog
2019-07-25 Juan Martinez <jjm@usebox.net>

 * release 0.21:

  - Add Swift storage policies support
  - Allow providing user/group in uid/gid options
  - Fix in SCP upload

2019-07-12 Juan Martinez <jjm@usebox.net>

 * release 0.20:

  - Keystone v3 support thanks to Julien Lutran

2016-11-03 Juan Martinez <juan@memset.com>

 * release 0.19:

  - Fix in recursive SCP
  - New functionally to allow setting the server ident

Thanks to Michel Nederlof for both patches.

2016-05-12 Juan Martinez <juan@memset.com>

 * release 0.18.1:

  - Minor documentation changes.

  paramiko 1.17.0+ is highly recommended as it adds support for
  hmac-sha2-256 digest.

2015-09-11 Juan Martinez <juan@memset.com>

 * release 0.18:

  - Configuration support to set SSH ciphers and digests.
  See sftpcloudfs.conf.example for further details.

  Thanks to Oliver Bristow for the suggested patch.

2015-01-16 Juan Martinez <juan@memset.com>

 * release 0.17:

  - Exposed SSH "keepalive" support with a new configuration token.

2015-01-15 Juan Martinez <juan@memset.com>

 * release 0.16:

  - pid file is now optional and it fixes inconsistent behaviour. It now
  will be used be used both in daemon and foreground mode as long as it
  is specified either in the CLI or in the configuration file.

  Thanks to Koert van der Veer for his report.

2014-10-06 Juan Martinez <juan@memset.com>

 * release 0.15:

  - Added "insecure" configuration token to support connections to auth
  servers with invalid SSL certificates.

This release requies ftpcloudfs >= 0.30 and uses swiftclient >= 2.0.0 (and
requests library). Please test thoroughly before upgrading your production
systems.

2014-05-15 Juan Martinez <juan@memset.com>

 * release 0.14.1:

  - Explicitly close connections. In some conditions Python garbage
  collector may not close all the opened connections. This is really uncommon,
  but users are encouraged to upgrade.

2014-04-17 Juan Martinez <juan@memset.com>

 * release 0.14:

  - New configuration token 'auth-timeout' to make auth time limit
    configurable (defaults to 60 seconds).
  - Added optional SSH negotiation time limit, exposed via 'negotiation-timeout'
    configuration token.
  - Clarification in sftpcloudfs.conf.example (re: quotes in string tokens).

  New timeout configuration can improve server's behaviour vs SSH brute-force
  attacks.

2014-04-10 Juan Martinez <juan@memset.com>

 * release 0.13.1:

  - ftp-cloudfs version bumped to 0.26.1 as it fixes a security bug.

2014-03-20 Juan Martinez <juan@memset.com>

 * release 0.13:

  - ftp-cloudfs version bumped to 0.26.

  - new hide-part-dir configuration token to hide ".part" files from
    large files support.
  - ftp-cloudfs 0.26 includes improved large file support (delete, rename)
  - copyright year bump.

2013-12-01 Juan Martinez <juan@memset.com>

 * release 0.12.1, 0.12.2:

  - ftp-cloudfs version bumped to 0.25.2.

  0.25.2 fixes an information leak vulnerability in sftp-cloudfs when the
  shared cache (memcache) is used. Under certain conditions the SFTP server
  could serve the cached directory listing of a different user.
  
  Upgrading ftp-cloudfs is highly recommended.

2013-11-15 Juan Martinez <juan@memset.com>

 * release 0.12:

  - Large object support, see 'split-large-files' configuration token.
    Requires ftp-cloudfs >= 0.25.

  For further info please read:
    
  http://docs.openstack.org/developer/swift/overview_large_objects.html

2013-10-18 Juan Martinez <juan@memset.com>

 * release 0.11:

  - Improved error logging.

2013-09-30 Juan Martinez <juan@memset.com>

 * release 0.10.1:

  - Small fixes for the SCP support.

2013-09-27 Juan Martinez <juan@memset.com>

 * release 0.10:

  - SCP support contributed by Koert van der Veer, thanks a lot!
  - New flag to enable/disable SCP in case you don't want to use that
    functionality (it will be enabled by default).

2013-08-29 Juan Martinez <juan@memset.com>

 * release 0.9:

  - Ported to use python-swiftclient instead of python-cloudfiles.
  - ftp-cloudfs 0.23+ is required

  From now on the project will focus on OpenStack Object Storage (swift);
  although Rackspace Cloud Files is still compatible.

  If you want to keep using python-cloudfiles, please use 0.8.1 with ftp-cloudfs 0.22.

2013-07-04 Juan Martinez <juan@memset.com>

 * release 0.8.1:

  - Minor release to fix a bug that was reducing the efficiency of the
    shared cache
  - Reverted paramiko transport log to CRITICAL

IMPORTANT: this is the last version supporting python-cloufiles.

2013-06-12 Juan Martinez <juan@memset.com>

 * release 0.8:

  - updated to work with ftp-cloudfs >= 0.21
  - increased paramiko transport log from CRITICAL to ERROR

2013-02-07 Juan Martinez <juan@memset.com>

 * release 0.7:

  - Auth 2.0 (keystone) support, requires ftp-cloudfs >= 0.20 and
    python-keystoneclient >= 0.1.3.66
  - Fixed a bug in the alternative configuration file handling with
    --config=FILE parameter

2013-02-05 Juan Martinez <juan@memset.com>

 * release 0.6.1:

  - Minor release to fix Debian packaging support
  - Copyright year bump

2012-10-30 Juan Martinez <juan@memset.com>

 * release 0.6:


debian/init.d
#!/bin/sh
### BEGIN INIT INFO
# Provides:          sftpcloudfs
# Required-Start:    $network $local_fs
# Required-Stop:
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: sftpcloudfs
# Description:       SFTP interface to Rackspace Cloud Files and OpenStack Object Storage.
### END INIT INFO

# Author: Christophe Le Guern <c35sys@gmail.com>

# PATH should only include /usr
PATH=/sbin:/usr/sbin:/bin:/usr/bin
DESC=sftpcloudfs                 # Introduce a short description here
NAME=sftpcloudfs                 # Introduce the short server's name here
CONFIGFILE=/etc/sftpcloudfs.conf # Configuration file
DAEMON=/usr/bin/sftpcloudfs      # Introduce the server's location here
LOGDIR=/var/log/${NAME}          # Logs directory
LOGFILE=${LOGDIR}/${NAME}.log    # Log file
DAEMON_ARGS="-l ${LOGFILE}"      # Arguments to run the daemon with
PIDDIR=/var/run/${NAME}
PIDFILE=${PIDDIR}/$NAME.pid
SCRIPTNAME=/etc/init.d/$NAME

# Exit if the package is not installed
[ -x $DAEMON ] || exit 0

# Exit if there is no configuration file
if [ ! -f ${CONFIGFILE} ]; then echo "Missing Configuration file"; exit 0; fi

# Read configuration variable file if it is present
[ -r /etc/default/$NAME ] && . /etc/default/$NAME

# Load the VERBOSE setting and other rcS variables
. /lib/init/vars.sh

# Define LSB log_* functions.
# Depend on lsb-base (>= 3.0-6) to ensure that this file is present.
. /lib/lsb/init-functions

#
# Function that starts the daemon/service
#
do_start()
{

	# create PID dir (/var/run/ is tmpfs)
	if [ ! -d ${PIDDIR} ]
	then
		mkdir -m 0700 -p ${PIDDIR}
		chown -R sftpcloudfs:sftpcloudfs ${PIDDIR}
	fi
	# Return
	#   0 if daemon has been started
	#   1 if daemon was already running
	#   2 if daemon could not be started
	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON --test > /dev/null \
		|| return 1
	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON -- \
		$DAEMON_ARGS \
		|| return 2
	# Add code here, if necessary, that waits for the process to be ready
	# to handle requests from services started subsequently which depend
	# on this one.  As a last resort, sleep for some time.
}

#
# Function that stops the daemon/service
#
do_stop()
{
	# Return
	#   0 if daemon has been stopped
	#   1 if daemon was already stopped
	#   2 if daemon could not be stopped
	#   other if a failure occurred
	start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --pidfile $PIDFILE --name $(basename $DAEMON)
	RETVAL="$?"
	[ "$RETVAL" = 2 ] && return 2
	# Wait for children to finish too if this is a daemon that forks
	# and if the daemon is only ever run from this initscript.
	# If the above conditions are not satisfied then add some other code
	# that waits for the process to drop all resources that could be
	# needed by services started subsequently.  A last resort is to
	# sleep for some time.
	start-stop-daemon --stop --quiet --oknodo --retry=0/30/KILL/5 --exec $DAEMON
	[ "$?" = 2 ] && return 2
	# Many daemons don't delete their pidfiles when they exit.
	rm -f $PIDFILE
	return "$RETVAL"
}

#
# Function that sends a SIGHUP to the daemon/service
#
do_reload() {
	#
	# If the daemon can reload its configuration without
	# restarting (for example, when it is sent a SIGHUP),
	# then implement that here.
	#
	start-stop-daemon --stop --signal 1 --quiet --pidfile $PIDFILE --name $(basename $DAEMON)
	return 0
}

case "$1" in
  start)
    [ "$VERBOSE" != no ] && log_daemon_msg "Starting $DESC " "$NAME"
    do_start
    case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
  ;;
  stop)
	[ "$VERBOSE" != no ] && log_daemon_msg "Stopping $DESC" "$NAME"
	do_stop
	case "$?" in
		0|1) [ "$VERBOSE" != no ] && log_end_msg 0 ;;
		2) [ "$VERBOSE" != no ] && log_end_msg 1 ;;
	esac
	;;
  status)
       status_of_proc "$DAEMON" "$NAME" && exit 0 || exit $?
       ;;
  #reload|force-reload)
	#
	# If do_reload() is not implemented then leave this commented out
	# and leave 'force-reload' as an alias for 'restart'.
	#
	#log_daemon_msg "Reloading $DESC" "$NAME"
	#do_reload
	#log_end_msg $?
	#;;
  restart|force-reload)
	#
	# If the "reload" option is implemented then remove the
	# 'force-reload' alias
	#
	log_daemon_msg "Restarting $DESC" "$NAME"
	do_stop
	case "$?" in
	  0|1)
		do_start
		case "$?" in
			0) log_end_msg 0 ;;
			1) log_end_msg 1 ;; # Old process is still running
			*) log_end_msg 1 ;; # Failed to start
		esac
		;;
	  *)
	  	# Failed to stop
		log_end_msg 1
		;;
	esac
	;;
  *)
	#echo "Usage: $SCRIPTNAME {start|stop|restart|reload|force-reload}" >&2
	echo "Usage: $SCRIPTNAME {start|stop|status|restart|force-reload}" >&2
	exit 3
	;;
esac

:


sftpcloudfs/main.py
#!/usr/bin/python
"""
Main function to setup the daemon process.

Copyright (C) 2011-2019 by Memset Ltd. http://www.memset.com/

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

"""

import os
import pwd
import signal
import sys
import logging
from logging.handlers import SysLogHandler
from ConfigParser import RawConfigParser, ParsingError
from optparse import OptionParser
import daemon
from Crypto import Random
import paramiko
from sftpcloudfs.server import ObjectStorageSFTPServer
from sftpcloudfs.constants import version, project_url, config_file, default_ks_service_type, \
    default_ks_tenant_separator, default_ks_endpoint_type
from ftpcloudfs.fs import ObjectStorageFS

class PIDFile(object):
    """
    PID file implementation using a context manager.

    Entering the context acquires the lock, raising OSError if the file exists
    and it's already locked. Leaving the context cleans the PID file.

    Some methods are implemented for compatibility with lockfile and python-daemon.
    """
    def __init__(self, pidfile=None):
        self.pidfile = pidfile
        self._fd = None

    def __enter__(self):
        return self.acquire()

    def __exit__(self, *args, **kwargs):
        self.release()

    def acquire(self):
        pid = os.getpid()
        fd = os.open(self.pidfile, (os.O_CREAT|os.O_EXCL|os.O_WRONLY), 0644)
        self._fd = os.fdopen(fd, "w")
        self._fd.write("%s\n" % pid)
        self._fd.flush()
        return pid

    def is_locked(self):
        try:
            os.stat(self.pidfile)
        except OSError:
            return False
        else:
            return True

    def i_am_locking(self):
        return self._fd != None

    def release(self):
        if self._fd:
            self._fd.close()
            os.remove(self.pidfile)
            self._fd = None

class Main(object):
    def __init__(self):
        """Parse configuration and CLI options."""
        global config_file

        # look for an alternative configuration file
        alt_config_file = False
        # used to show errors before we actually start parsing stuff
        parser = OptionParser()
        for arg in sys.argv:
            if arg == '--config':
                try:
                    alt_config_file = sys.argv[sys.argv.index(arg)+1]
                    config_file = alt_config_file
                except IndexError:
                    pass
            elif arg.startswith('--config='):
                _, alt_config_file = arg.split('=', 1)
                if alt_config_file == '':
                    parser.error("--config option requires an argument")
                config_file = alt_config_file

        config = RawConfigParser({'auth-url': None,
                                  'insecure': False,
                                  'host-key-file': None,
                                  'bind-address': "127.0.0.1",
                                  'port': 8022,
                                  'server-ident': 'sftpcloudfs_%s' % version,
                                  'memcache': None,
                                  'max-children': "20",
                                  'auth-timeout': "60",
                                  'negotiation-timeout': "0",
                                  'keepalive': "0",
                                  'ciphers': None,
                                  'digests': None,
                                  'log-file': None,
                                  'syslog': 'no',
                                  'verbose': 'no',
                                  'scp-support': 'yes',
                                  'pid-file': None,
                                  'uid': None,
                                  'gid': None,
                                  'split-large-files': "0",
                                  'hide-part-dir': "no",
                                  # keystone auth support
                                  'keystone-auth': False,
                                  'keystone-auth-version': '2.0',
                                  'keystone-region-name': None,
                                  'keystone-tenant-separator': default_ks_tenant_separator,
                                  'keystone-domain-separator': '@',
                                  'keystone-service-type': default_ks_service_type,
                                  'keystone-endpoint-type': default_ks_endpoint_type,
                                  'storage-policy': None,
                                  })

        try:
            if not config.read(config_file) and alt_config_file:
                # the default conf file is optional
                parser.error("failed to read %s" % config_file)
        except ParsingError as ex:
             parser.error("failed to read %s: %s" % (config_file, ex.message))

        if not config.has_section('sftpcloudfs'):
            config.add_section('sftpcloudfs')

        parser = OptionParser(version="%prog " + version,
                              description="This is a SFTP interface to OpenStack " + \
                                    "Object Storage (Swift).",
                              epilog="Contact and support at: %s" % project_url)

        parser.add_option("-a", "--auth-url", dest="authurl",
                          default=config.get('sftpcloudfs', 'auth-url'),
                          help="Authentication URL")

        parser.add_option("--insecure", dest="insecure",
                          action="store_true",
                          default=config.get('sftpcloudfs', 'insecure'),
                          help="Allow to access servers without checking SSL certs")

        parser.add_option("-k", "--host-key-file", dest="host_key",
                          default=config.get('sftpcloudfs', 'host-key-file'),
                          help="Host RSA key used by the server")

        parser.add_option("-b", "--bind-address", dest="bind_address",
                          default=config.get('sftpcloudfs', 'bind-address'),
                          help="Address to bind (default: 127.0.0.1)")

        parser.add_option("-p", "--port", dest="port",
                          type="int",
                          default=config.get('sftpcloudfs', 'port'),
                          help="Port to bind (default: 8022)")

        parser.add_option("--server-ident", dest="server_ident",
                          type="str",
                          default=config.get('sftpcloudfs', 'server-ident'),
                          help="Server ident to use when sending the SSH banner to the " + \
                                  "client (default: sftpcloudfs_%s)" % version)

        memcache = config.get('sftpcloudfs', 'memcache')
        if memcache:
            memcache = [x.strip() for x in memcache.split(',')]
        parser.add_option('--memcache',
                          type="str",
                          dest="memcache",
                          action="append",
                          default=memcache,
                          help="Memcache server(s) to be used for cache (ip:port)")


        parser.add_option("-l", "--log-file", dest="log_file",
                          default=config.get('sftpcloudfs', 'log-file'),
                          help="Log into provided file")

        parser.add_option("-f", "--foreground", dest="foreground",
                          action="store_true",
                          default=False,
                          help="Run in the foreground (don't detach from terminal)")

        parser.add_option("--disable-scp", dest="no_scp",
                          action="store_true",
                          default=not config.getboolean('sftpcloudfs', 'scp-support'),
                          help="Disable SCP support (default: enabled)")

        parser.add_option("--syslog", dest="syslog",
                          action="store_true",
                          default=config.getboolean('sftpcloudfs', 'syslog'),
                          help="Enable logging to system logger (daemon facility)")

        parser.add_option("-v", "--verbose", dest="verbose",
                          action="store_true",
                          default=config.getboolean('sftpcloudfs', 'verbose'),
                          help="Show detailed information on logging")

        parser.add_option('--pid-file',
                          type="str",
                          dest="pid_file",
                          default=config.get('sftpcloudfs', 'pid-file'),
                          help="Full path to the pid file location")

        parser.add_option('--uid',
                          dest="uid",
                          default=config.get('sftpcloudfs', 'uid'),
                          help="UID to drop the privileges to when in daemon mode")

        parser.add_option('--gid',
                          dest="gid",
                          default=config.get('sftpcloudfs', 'gid'),
                          help="GID to drop the privileges to when in daemon mode")

        parser.add_option('--keystone-auth',
                          action="store_true",
                          dest="keystone",
                          default=config.get('sftpcloudfs', 'keystone-auth'),
                          help="Use Keystone auth (requires keystoneclient)")

        parser.add_option('--keystone-auth-version',
                          type="str",
                          dest="auth_version",
                          default=config.get('sftpcloudfs', 'keystone-auth-version'),
                          help="Identity API version to be used (default: 2.0)")

        parser.add_option('--keystone-region-name',
                          type="str",
                          dest="region_name",
                          default=config.get('sftpcloudfs', 'keystone-region-name'),
                          help="Region name to be used in Keystone auth")

        parser.add_option('--keystone-tenant-separator',
                          type="str",
                          dest="tenant_separator",
                          default=config.get('sftpcloudfs', 'keystone-tenant-separator'),
                          help="Character used to separate tenant_name/username in Keystone auth, " + \
                              "default: TENANT%sUSERNAME" % default_ks_tenant_separator)

        parser.add_option('--keystone-domain-separator',
                          type="str",
                          dest="domain_separator",
                          default=config.get('sftpcloudfs', 'keystone-domain-separator'),
                          help="Character used to separate project_name/project_domain_name " + \
                               "and username/user_domain_name in Keystone auth v3 (default: @)")

        parser.add_option('--keystone-service-type',
                          type="str",
                          dest="service_type",
                          default=config.get('sftpcloudfs', 'keystone-service-type'),
                          help="Service type to be used in Keystone auth, default: %s" % default_ks_service_type)

        parser.add_option('--keystone-endpoint-type',
                          type="str",
                          dest="endpoint_type",
                          default=config.get('sftpcloudfs', 'keystone-endpoint-type'),
                          help="Endpoint type to be used in Keystone auth, default: %s" % default_ks_endpoint_type)

        parser.add_option('--config',
                          type="str",
                          dest="config",
                          default=config_file,
                          help="Use an alternative configuration file")

        parser.add_option("--storage-policy",
                          type="str",
                          dest="storage_policy",
                          default=config.get('sftpcloudfs', 'storage-policy'),
                          help="Swift storage policy to be used")

        (options, args) = parser.parse_args()

        # required parameters
        if not options.authurl:
            parser.error("No auth-url provided")

        if not options.host_key:
            parser.error("No host-key-file provided")

        try:
            self.host_key = paramiko.RSAKey(filename=options.host_key)
        except (IOError, paramiko.SSHException), e:
            parser.error("host-key-file: %s" % e)

        if options.memcache:
            ObjectStorageFS.memcache_hosts = options.memcache
            try:
                ObjectStorageFS(None, None, None)
            except (ValueError, TypeError):
                parser.error("memcache: invalid server address, ip:port expected")

        if options.pid_file:
            self.pidfile = PIDFile(options.pid_file)
            if self.pidfile.is_locked():
                parser.error("pid-file found: %s\nIs the server already running?" % options.pid_file)
        else:
            self.pidfile = None

        try:
            options.max_children = int(config.get('sftpcloudfs', 'max-children'))
        except ValueError:
            parser.error('max-children: invalid value, integer expected')

        try:
            options.auth_timeout = int(config.get('sftpcloudfs', 'auth-timeout'))
        except ValueError:
            parser.error('auth-timeout: invalid value, integer expected')

        if options.auth_timeout <= 0:
            parser.error('auth-timeout: invalid value')

        try:
            options.negotiation_timeout = int(config.get('sftpcloudfs', 'negotiation-timeout'))
        except ValueError:
            parser.error('negotiation-timeout: invalid value, integer expected')

        if options.negotiation_timeout < 0:
            parser.error('negotiation-timeout: invalid value')

        try:
            options.keepalive = int(config.get('sftpcloudfs', 'keepalive'))
        except ValueError:
            parser.error('keepalive: invalid value, integer expected')

        if options.keepalive < 0:
            parser.error('keepalive: invalid value')

        options.secopts = {}
        ciphers = config.get('sftpcloudfs', 'ciphers')
        if ciphers:
            options.secopts["ciphers"] = [x.strip() for x in ciphers.split(',')]

        digests = config.get('sftpcloudfs', 'digests')
        if digests:
            options.secopts["digests"] = [x.strip() for x in digests.split(',')]

        try:
            options.split_size = int(config.get('sftpcloudfs', 'split-large-files'))*10**6
        except ValueError:
            parser.error('split-large-files: invalid size, integer expected')

        options.hide_part_dir = config.getboolean('sftpcloudfs', 'hide-part-dir')

        if options.keystone:
            keystone_keys = ('auth_version', 'region_name', 'tenant_separator', 'domain_separator', 'service_type', 'endpoint_type')
            options.keystone = dict((key, getattr(options, key)) for key in keystone_keys)

        if options.uid:
            try:
                options.uid = int(options.uid)
            except ValueError:
                try:
                    options.uid = pwd.getpwnam(options.uid).pw_uid
                except KeyError:
                    parser.error("uid: Invalid uid: %s" % options.uid)

        if options.gid:
            try:
                options.gid = int(options.gid)
            except ValueError:
                try:
                    options.gid = pwd.getpwnam(options.gid).pw_gid
                except KeyError:
                    parser.error("gid: Invalid gid: %s" % options.gid)

        self.options = options

    def setup_log(self):
        """Setup server logging facility."""
        self.log = paramiko.util.get_logger("paramiko")

        if self.options.log_file:
            paramiko.util.log_to_file(self.options.log_file)

        if self.options.syslog is True:
            try:
                handler = SysLogHandler(address='/dev/log',
                                        facility=SysLogHandler.LOG_DAEMON)
            except IOError:
                # fall back to UDP
                handler = SysLogHandler(facility=SysLogHandler.LOG_DAEMON)
            finally:
                handler.setFormatter(logging.Formatter('%(name)s[%(_threadid)s]: %(levelname)s: %(message)s'))
                self.log.addHandler(handler)

        if self.options.foreground:
            handler = logging.StreamHandler()
            handler.setFormatter(logging.Formatter('%(asctime)s: %(name)s[%(_threadid)s]: %(levelname)s: %(message)s'))
            self.log.addHandler(handler)

        if self.options.verbose:
            # enable debug for the root logger (used by ObjectStorageFS)
            logging.getLogger().setLevel(logging.DEBUG)
            self.log.setLevel(logging.DEBUG)
            self.log.debug(self.options)
        else:
            self.log.setLevel(logging.INFO)


    def run(self):
        """Run the server."""
        server = ObjectStorageSFTPServer((self.options.bind_address, self.options.port),
                                          host_key=self.host_key,
                                          authurl=self.options.authurl,
                                          max_children=self.options.max_children,
                                          keystone=self.options.keystone,
                                          no_scp=self.options.no_scp,
                                          split_size=self.options.split_size,
                                          hide_part_dir=self.options.hide_part_dir,
                                          auth_timeout=self.options.auth_timeout,
                                          negotiation_timeout=self.options.negotiation_timeout,
                                          keepalive=self.options.keepalive,
                                          insecure=self.options.insecure,
                                          secopts=self.options.secopts,
                                          server_ident=self.options.server_ident,
                                          storage_policy=self.options.storage_policy,
                                          )

        dc = daemon.DaemonContext()
        dc.pidfile = self.pidfile

        if self.options.uid:
            dc.uid = self.options.uid

        if self.options.gid:
            dc.gid = self.options.gid

        # FIXME: we don't know the fileno for Random open files, but they're  < 16
        dc.files_preserve = range(server.fileno(), 16)

        if self.options.foreground:
            dc.detach_process = False
            dc.stderr = sys.stderr

        with dc:
            Random.atfork()
            self.setup_log()
            try:
                if os.getuid() == 0:
                    self.log.warning("UID is 0, running as root is not recommended")

                self.log.info("Listening on %s:%s" % (self.options.bind_address, self.options.port))
                server.serve_forever()
            except (SystemExit, KeyboardInterrupt):
                self.log.info("Terminating...")
                if server.active_children:
                    for pid in server.active_children:
                        os.kill(pid, signal.SIGTERM)
                server.server_close()

        if self.pidfile and self.pidfile.i_am_locking():
            self.pidfile.release()

        return 0

